allegro/nn/_strided/_cueq_contracter.py

```diff
@@ -0,0 +1,114 @@
+# This file is a part of the `allegro` package. Please see LICENSE and README at the root for information on using it.
+import torch
+
+import cuequivariance as cue
+import cuequivariance_torch as cuet
+
+from nequip.nn import scatter
+from ._contract import Contracter
+
+import itertools
+from typing import Dict
+
+
+def allegro_tp_desc(
+    irreps1: cue.Irreps,
+    irreps2: cue.Irreps,
+    irreps3: cue.Irreps,
+) -> cue.EquivariantPolynomial:
+    """Construct the Allegro version of channelwise tensor product descriptor.
+
+    subscripts: ``weights[u],lhs[iu],rhs[ju],output[ku]``
+
+    Args:
+        irreps1 (Irreps): Irreps of the first operand.
+        irreps2 (Irreps): Irreps of the second operand.
+        irreps3 (Irreps): Irreps of the output to consider.
+    """
+    common_mul = irreps1[0].mul
+
+    d = cue.SegmentedTensorProduct.from_subscripts("u,iu,ju,ku+ijk")
+
+    for mul, ir in irreps1:
+        assert mul == common_mul
+        d.add_segment(1, (ir.dim, mul))
+    for mul, ir in irreps2:
+        assert mul == common_mul
+        d.add_segment(2, (ir.dim, mul))
+    for mul, ir in irreps3:
+        d.add_segment(3, (ir.dim, common_mul))
+
+    for (i3, (mul3, ir3)), (i1, (mul1, ir1)), (i2, (mul2, ir2)) in itertools.product(
+        enumerate(irreps3), enumerate(irreps1), enumerate(irreps2)
+    ):
+        if ir3 in ir1 * ir2:
+            for cg in cue.clebsch_gordan(ir1, ir2, ir3):
+                d.add_path(None, i1, i2, i3, c=cg)
+
+    return cue.EquivariantPolynomial(
+        [
+            cue.IrrepsAndLayout(irreps1.new_scalars(d.operands[0].size), cue.ir_mul),
+            cue.IrrepsAndLayout(irreps1, cue.ir_mul),
+            cue.IrrepsAndLayout(irreps2, cue.ir_mul),
+        ],
+        [cue.IrrepsAndLayout(irreps3, cue.ir_mul)],
+        cue.SegmentedPolynomial.eval_last_operand(d),
+    )
+
+
+class CuEquivarianceContracter(Contracter):
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        ir1 = cue.Irreps("O3", [(self.mul, ir) for _, ir in self.irreps_in1])
+        ir2 = cue.Irreps("O3", [(self.mul, ir) for _, ir in self.irreps_in2])
+        irout = cue.Irreps("O3", [(self.mul, ir) for _, ir in self.irreps_out])
+        self.cuet_sp = cuet.SegmentedPolynomial(
+            allegro_tp_desc(ir1, ir2, irout).polynomial.flatten_coefficient_modes(),
+            math_dtype=self.w3j.dtype,
```

Comment:
This ends up being model_dtype?

Reply:
yes, just using it because it's the only buffer that will always be there